lib/husky_musher/app.py

```diff
@@ -152,3 +156,74 @@ def main():
 def set_cache_control(response):
     response.headers["Cache-Control"] = "no-store"
     return response
+
+
+def extract_user_info(environ: dict) -> Dict[str, str]:
+    """
+    Extracts attributes of the authenticated user, provided by UW's IdP via our
+    Shibboleth SP, from the request environment *environ*.
+
+    Keys of the returned dict match those used by our REDCap project.
+    """
+    return {
+        "netid": f"{environ['uid']}",
+
+        # This won't always be @uw.edu.
+        "email": environ.get("mail", ""),
+
+        # Given name will include any middle initial/name.  Both name fields
+        # will contain the preferred name parts, if set, otherwise the
+        # administrative name parts.
+        "core_participant_first_name": environ.get("givenName", ""),
+        "core_participant_last_name":  environ.get("surname", ""),
+
+        # Department is generally a colon-separated set of
+        # increasingly-specific labels, starting with the School.
+        "uw_school": environ.get("department", ""),
+
+        **extract_affiliation(environ),
+    }
+
+
+def extract_affiliation(environ: dict) -> Dict[str, str]:
+    """
+    Transforms a multi-value affiliation string into our REDCap fields.
+
+    Keys of the returned dict match those used by our REDCap project.
+
+    >>> extract_affiliation({"unscoped-affiliation": "member;faculty;employee;alum"})
+    {'affiliation': 'faculty', 'affiliation_other': ''}
+
+    >>> extract_affiliation({"unscoped-affiliation": "member;student;staff"})
+    {'affiliation': 'student', 'affiliation_other': ''}
+
+    >>> extract_affiliation({"unscoped-affiliation": "member;staff;alum"})
+    {'affiliation': 'staff', 'affiliation_other': ''}
+
+    >>> extract_affiliation({"unscoped-affiliation": "member;employee"})
+    {'affiliation': 'staff', 'affiliation_other': ''}
+
+    >>> extract_affiliation({"unscoped-affiliation": "member;affiliate;alum"})
+    {'affiliation': 'other', 'affiliation_other': 'affiliate;alum'}
+
+    >>> extract_affiliation({"unscoped-affiliation": "member"})
+    {'affiliation': '', 'affiliation_other': ''}
+
+    >>> extract_affiliation({})
+    {'affiliation': '', 'affiliation_other': ''}
+    """
+    raw_affilations = environ.get("unscoped-affiliation", "")
+
+    # "Member" is uninteresting and uninformative; a generic catch-all.
+    # The empty string might arise from our fallback above.
+    affiliations = set(raw_affilations.split(";")) - {"member",""}
+
+    rules = [
+        ("faculty"  in affiliations,    {"affiliation": "faculty",  "affiliation_other": ""}),
+        ("student"  in affiliations,    {"affiliation": "student",  "affiliation_other": ""}),
```

Comment:
In REDCap, the "student" affiliation takes precedence (Are you a student, faculty, staff, or volunteer at the University of Washington? If you are both working and enrolled in classes at the University of Washington, please select "Student".) 

If I'm understanding this function correctly, should the `"student" in affiliations` condition be the first in the list?

Reply:
Thanks! Fixed in a new commit.